### ISSUE TITLE
Reports memory leak

### ISSUE THREAD

**Issue by RachL** [openfoodfoundation/openfoodnetwork#10016]
## Description

We can reproduce the memory leak from reports: let's try to fix it!

Copying some ideas from slack: 

From Lynne : 

I'd be starting investigation looking at [this line](https://github.com/openfoodfoundation/openfoodnetwork/blob/0f618dfbf9b6250ecfab988cef5e4329581575dd/lib/reporting/reports/order_cycle_management/base.rb#L32) in the Order Cycle reports base controller. Customer balances are processing intensive. I'm just realising now that line was added for correctness during the reports refactorring. In the past we didn't do full customer balances in these reports due to performance overhead.
The balance will only be correct in certain circumstances on this report - for only the most recent order per customer returned. Hence it might not be the most useful field. Something to consider.

and

Maikel suggested trying to see if there’s any memory management config that could help (in Ruby, or maybe even rails or puma..) 

## Steps to Reproduce

1. Take order cycle customer total report for example
2. Try to launch it for a period of 3 months
3. See the memory is pilling up without releasing

## Workaround

Restart Puma. See https://github.com/openfoodfoundation/openfoodnetwork/issues/9859#issuecomment-1316126889

## Severity
bug-s3: a feature is broken but there is a workaround

## Your Environment

* Version used: v4.2.21.1
* Browser name and version: Al
* Operating System and version (desktop or mobile): All

**Comment by dacook** [comment]
Good find! Currently my suspicion is a memory leak due to the usage of procs in reports, because it will spike up due to a large report, and stay up. It doesn't return down even after days. The pattern is not perfectly clear but I think that's what's happening. 

However I think this is a good point:

> Instead of chasing unknown culprits, we went straight to the common source of many of our challenges: known issues with Ruby. We began with jemalloc, thinking, “If this solves the problem, we won’t need to investigate for leaks and local caching.”

It seems like it wouldn't take much effort to try this out, so perhaps we should try it anyway. We can prove the concept on a staging or vagrant VM, and if it happens to be successful, update ofn-install and roll out to all servers.

**Comment by dacook** [comment]
Sharing my notes from earlier tests:

## Order Cycle Sales Report  -  2021 Jan-Dec (CSV)
Memory usage grows, and does not return. On subsequent execution memory usage grows further. Continued running the same report again and again on staging envs until the process died. (first time for uk/ca, took about 6 times on FR staging before it died)

- Local Dev (FR)  Success Memory usage +0.3gb
- FR Staging (FR) Success Memory usage +1.1gb
- UK Staging (CA) Failed  Memory usage +0.5gb

### FR Prod
I tried to run the same report on FR Prod, and didn’t notice much slowdown on the site, but the memory leak is clear (a process went from 1gb to 5gb, and remains).

No use of swap, but looking back at datadog I can see that's probably what's happening when it continues growing. Can also see that the memory isn't unallocated even after several days.

I also noticed full utilisation of a CPU core. On this server because there are 8 cores, and only 4 puma processes. Most other servers only have 4 cores and 2 puma processes. This could be a contributor but I don’t know.

(After this test, the puma server had to be restarted!)

**Comment by abdellani** [comment]
Hi @dacook 

Thank you for sharing your notes :pray:.

Here is a summary of notes:
1. When I started playing with ansible to install jemalloc, I noticed the compile instructions that you mentioned in your [comment](https://github.com/openfoodfoundation/ofn-install/pull/850#issuecomment-1354064292).

I used the [lsof command](https://superuser.com/a/310205) to check if the running puma servers are using `jemalloc`.
Here are the results from the FR staging server
```bash
openfoodnetwork@123:~$ ps aux |grep puma
openfoo+ 13393  0.0  0.5 134252 22356 ?        Ss   06:08   0:01 puma 5.6.5 (unix:///home/openfoodnetwork/apps/openfoodnetwork/shared/sock/puma.openfoodnetwork.sock) [2022-12-16-031031]
openfoo+ 13514  0.1  6.4 1580400 255496 ?      Sl   06:08   0:10 puma: cluster worker 0: 13393 [2022-12-16-031031]
openfoo+ 13522  0.0  5.9 1565924 235200 ?      Sl   06:08   0:02 puma: cluster worker 1: 13393 [2022-12-16-031031]
openfoo+ 13528  0.0  5.4 1225908 214560 ?      Sl   06:08   0:00 puma: cluster worker 2: 13393 [2022-12-16-031031]
openfoo+ 13534  0.0  6.0 1565924 240248 ?      Sl   06:08   0:02 puma: cluster worker 3: 13393 [2022-12-16-031031]
openfoo+ 15223  0.0  0.0  17388  1056 pts/4    S+   07:36   0:00 grep puma

openfoodnetwork@123:~$ lsof -p 13514|grep jemalloc
ruby    13514 openfoodnetwork  mem       REG                8,1   215656    16149 /usr/lib/x86_64-linux-gnu/libjemalloc.so.1
```
From the second command, I can say that `jemalloc` is loaded by puma.
I checked UK staging and concluded the same thing.
@filipefurtad0 check UK Prod and got the same results from `lsof`.

I double-checked with [this method](https://stackoverflow.com/a/57303366/12818968)
```ruby
irb(main):001:0> RbConfig::CONFIG['MAINLIBS']
=> "-lz -lpthread -lrt -lrt -ljemalloc -ldl -lcrypt -lm "
```
I think it's safe to say that `jemalloc` is in use.

2. After this conclusion, I decided to do some research about the GC (Garbage Collector).
I used to think that the GC immediately releases the memory space allocated for an object that's not in use.

I started by reading this [article](https://www.speedshop.co/2017/12/04/malloc-doubles-ruby-memory.html)

- I opened the FR staging
- restarted puma
- run GC.stat
- run a report from the admin dashboard
- run GC.stat 
- wait few minutes, and run run GC.stat

I used the following formula from [article](https://www.speedshop.co/2017/12/04/malloc-doubles-ruby-memory.html) to calculate the  % of the occupied eden pages.

```
# live_slots / (eden_pages * slots_per_page)
```
![image](https://user-images.githubusercontent.com/16113997/208103735-1e7991cd-9a20-46aa-8c1f-70afd51c5807.png)

>96% looks like a good percentage, but does this mean that our app doesn't suffer from a memory fragmentation issue, and jemalloc is doing good job? I honestly can't answer.


3. I moved to the following [article](https://www.speedshop.co/2017/03/09/a-guide-to-gc-stat.html) which explaing the attribtues returned by the GC.stat

The following paragraph is interesting: 
```
As for practical use of the slot counts in GC.stat, I’d suggest monitoring heap_free_slots if you’re having memory bloat issues. Large numbers of free slots (like more than 300,000) usually indicates that you have a few controller actions which are allocating large numbers of objects and then freeing them, which can permanently bloat the size of your Ruby process. For more about fixing that, check out my Rubyconf talk of memory issues in Ruby.
``` 
From my records:
![image](https://user-images.githubusercontent.com/16113997/208108908-924b881d-67e3-470d-a10e-36f9c840d442.png)

The records are much less than 300,000 but I did the test in a small system with less than 4GB of RAM.

This link in the paragraph is broken, [Here a new link on Youtube (RubyConf 2016 - Halve Your Memory Usage With These 12 Weird Tricks)](https://www.youtube.com/watch?v=kZcqyuPeDao&ab_channel=Confreaks). 

My final conclusion after watching the conference:


1. the garbage collector doesn't immediately release the memory to the system when the objects are not needed anymore. The algorithm is much more complex than that, There are trasholds and contraints that are taken into account.

I don't fully understand how it works, but I find the following image usefull to understand what going inside.
- The red line represents the memory space hold by the app.
- The blue line represents the memory in use.
![image](https://user-images.githubusercontent.com/16113997/208110319-785c20f3-abb3-4b6c-8977-a7c30a1d3507.png)

2. It's possible to play with GC parameters to change it's behaviour, but this is not recommended as it can create other problems.
3. At this level, I support [your suspicion](https://github.com/openfoodfoundation/openfoodnetwork/issues/10016#issuecomment-1344957557) and I think that way to solve the problem is through optimizing the reporting generator. 
Is this only solution? I don't know.

**Comment by abdellani** [comment]
Update:

I dumped one of the staging databases, and install [oink](https://github.com/noahd1/oink) 

> Oink adds memory and active record instantiation information to rails log during runtime and provides an executable to help digest the enhanced logs.

```ruby
---- MEMORY THRESHOLD ----
THRESHOLD: 75 MB

-- SUMMARY --
Worst Requests:
1. Dec 23 08:12:44, 327680 KB, admin/reports#show
2. Dec 22 11:16:45, 196608 KB, admin/reports#show
3. Dec 23 08:36:35, 165072 KB, admin/reports#show
4. Dec 22 11:08:41, 133124 KB, api/v0/products#bulk_products
5. Dec 23 08:14:48, 115736 KB, admin/reports#show
6. Dec 22 11:48:27, 83152 KB, spree/admin/overview#index
7. Dec 22 11:40:06, 79880 KB, admin/reports#show

Worst Actions:
5, admin/reports#show
1, api/v0/products#bulk_products
1, spree/admin/overview#index

Aggregated Totals:
Action                       	Max	Mean	Min	Total	Number of requests
admin/reports#show           	327680	176995	79880	884976	5
api/v0/products#bulk_products	133124	133124	133124	133124	1
spree/admin/overview#index   	83152	83152	83152	83152	1
```
```ruby
---- OINK FOR ACTIVERECORD ----
THRESHOLD: 500 Active Record objects per request

-- SUMMARY --
Worst Requests:
1. Dec 23 08:14:48, 119448, admin/reports#show
2. Dec 23 08:36:35, 119413, admin/reports#show
3. Dec 23 08:12:44, 119408, admin/reports#show
4. Dec 23 08:35:22, 119395, admin/reports#show
5. Dec 22 11:40:06, 21970, admin/reports#show
6. Dec 22 11:18:09, 21968, admin/reports#show
7. Dec 22 11:25:04, 21938, admin/reports#show
8. Dec 22 11:41:50, 21934, admin/reports#show
9. Dec 22 11:16:45, 20412, admin/reports#show
10. Dec 22 11:15:58, 18215, admin/reports#show

Worst Actions:
31, admin/reports#show
12, api/v0/products#overridable
7, admin/order_cycles#index
2, spree/admin/products#index
2, api/v0/products#bulk_products
2, spree/admin/orders#index
2, admin/variant_overrides#index
1, spree/admin/products#edit
1, admin/enterprises#visible
1, admin/reports#index

Aggregated Totals:
```
The worst report allocates ~327MB and creates  ~119K active record objects. 

I checked some articles  https://www.speedshop.co/blog/ and some chapters  [THE COMPLETE GUIDE TO RAILS PERFORMANCE](https://www.railsspeed.com/), here's what I found:

1. Avoid loading all the data at once. 
   a. Use `in_batches`, or `find_each` from Active Record. 
   b. Instead of loading 100k Active Record objects at once, we can load bulks of 10k instead, and populate the report.
   c. we'll need to run multiple database queries which can reduce the performance.
   d. All the examples that I have seen, use those methods in the context of updating records. I don't know if this is usable in our context
2. Use pluck to load the needed fields only and avoid instantiating AR objects. Not sure if this is possible with the reports that interact with multiple models.
3. if using pluck is not possible, we can reduce the size of the AR objects by using `select`. 
4. Use streams `ActionController::Streaming` to stream the report to the browser before it starts to allocate too much memory in the server. I don't know if this is usable in our case.

**Comment by dacook** [comment]
Hi @abdellani , thanks for thoroughly documenting your findings! 

So in summary:
* We're already using the best memory allocator (`jemalloc`)
* There's a chance adjusting the parameters for the Ruby garbage collector could improve performance, but it's not recommended and probably won't fix a leak anyway
* We could also optimise our reports framework to improve memory performance using the above tips, but that won't fix a leak

So, given we believe the problem is a leak, and that it could be caused by the procs,  the next thing I can think of is to try altering the test case to avoid procs and see if that makes a difference. I'll have a go at doing that...

**Comment by mkllnk** [comment]
A few more thoughts:

### Load data in smaller batches

This reduces the maximum used memory and would help in any case but I have two concerns. If we have a memory leak, it won't solve it. And I guess that some reports need to sort rows after loading which means that we do need all the data to process it. Sorting based on in-memory values is a huge problem in optimising performance but may simply be a requirement for reports. We need to check if we can sort the most complicated reports in the database.

### Possible memory leak

David's tests seem to confirm that we have a leak. Mohamed found lots of freed up slots though. We have to reconfirm this.

Procs can have circular dependencies though. There's a good chance that we have a memory leak.

**Comment by mkllnk** [comment]
I just did another experiment on fr-staging:
![Screenshot from 2023-01-04 14-31-49](https://user-images.githubusercontent.com/3524483/210479610-ee478516-9268-4348-93f7-3834822bef81.png)

I reduced the amount of puma workers to one so that the request is always handled by the same worker. Otherwise I would need to monitor the memory of four different workers. After a restart, the puma worker used 250 MB and each report generation added to the memory consumption. The report is fairly small and takes only a second to generate. So I could do a lot of iterations. The added memory was 35 MB to start with but then tapered off to 1 MB or less after ten requests.

I still find this result confusing. The logarithmic shape of the memory curve suggests that everything is fine. A worker just needs over 500 MB to deal with these report requests. Added memory consumption can be due to memory fragmentation. But growing memory needs through fragmentation or a memory leak are still a problem.

au-prod has only two workers and 12 GB of memory. One worker uses 1.1 GB of memory and that was restarted this morning. It would be good to check before the next deploy. We may just need to account for a lot of memory per worker.

One solution has been to add more workers to keep the site responsive while reports take a long time to generate. That made the memory problem worse. Especially because each worker will handle a big request at some point and will grow its memory needs.

There's no limit to the size of the reports. No matter how much memory we have, you can always generate bigger and bigger reports to crash the system. After the consumption on fr-staging tapered off, I extended the date range and the single worker used up all memory and died. We probably need to confine the report generation to a background job. Then the big memory footprint is only on one background process instead on a number of puma workers. And we can set limits around execution time and memory consumption.

Unfortunately, processing report data in batches is not possible because we need to evaluate and sort in memory.

I guess that there are is still scope for optimisation within the reports code which would make all problems smaller but can't solve the fact that some reports are huge.

My conclusion is to prioritise moving reports to a background job. We should also try to optimise the code.

**Comment by dacook** [comment]
That clarifies things more! 
However I'm still wondering why did fr-prod suffer ongoing performance issues? It would slowly get worse and worse until the only fix was to restart puma. 

My only guess is that because so much memory was allocated, the system was having to use page files for allocating memory, which caused the slowdown.
But I see that that since September, it hasn't ever used the maximum of system memory, so I wouldn't have thought it would be needing page files.
![Screen Shot 2023-01-04 at 3 10 00 pm](https://user-images.githubusercontent.com/4188088/210483976-014f1350-df3f-44a9-a0c2-e8f57f87e543.png)
This is the page file usage: barely anything.
![Screen Shot 2023-01-04 at 3 13 40 pm](https://user-images.githubusercontent.com/4188088/210484142-e69cc474-59b6-453e-840f-182b6529dd71.png)
So that doesn't explain it.

Perhaps it was due to the Garbage Collector having to work so hard to clean up and compact all the memory?

**Comment by mkllnk** [comment]
David and I have come up with a plan. As a quick solution we should further increase the server memory or reduce the amount of workers. You can always run bigger and bigger reports but most of the time the workers don't grow much bigger than 4GB in a week. So we can say that fr-prod needs at least `3 GB + 4x4 GB = 19 GB` for its 4 workers or we keep the current memory and reduce the worker count to 2. Australia is going quite well with 2 workers.

Next up is moving report generation into a background job. So if reports add a memory overhead of 3 GB, for example, then it would be contained in one background process instead of needing 3 GB per worker. It's also easier to limit resources in background jobs.

In parallel, we can try to optimise the code to use less resources. We still don't know exactly how, but there should be scope to do so.

How does that sound, @RachL?

**Comment by dacook** [comment]
Maikel and I discussed what would be required to defer report generation into a background job. 
Currently, users receive feedback directly (whether on-screen or as a downloaded file), usually quickly, but large reports can take a long time. The long reports mean that puma workers can't respond to other users, and can impact performance.

We have an idea that would not be too complicated to implement, and would still provide direct feedback for small reports.

1. User enters report parameters and clicks submit
2. A new Sidekiq job is created with these report parameters
3. The controller waits up to 20 seconds to complete. 
    - If completed, the user receives the results as usual
    - If still running, a message will appear: "this report is taking longer to process, it will be emailed to you once complete"
4. If the Sidekiq job completes after approx 20 seconds, it emails the report to the user (or a link to download).

We'll write up some more tech details tomorrow, and estimate this if it is deemed acceptable.

**Comment by RachL** [comment]
@mkllnk @dacook thanks for all these info! 

We've talked about it in product circle this morning. If a link to download was generated, this would have our favor. It would enable to display the link within the app. However if we do something like we currently have with Bulk Invoice Printing (where there is a background job but the user has to wait for the page to display the link - if he goes elsewhere, the link will never show up and he has to start again), this is not a good solution and an email with the link is preferable.

Could you estimate both version: 

1. Email the link
2. Display the link on the app: either in a dedicated menu or something like a notification (see https://github.com/openfoodfoundation/wishlist/issues/61 )

thanks :)

**Comment by mkllnk** [comment]
Our reasoning for sending an email was that the browser tab may be closed or the connection to the server may be disrupted and then the email is the only way to access the report. It's the most solid solution and therefore we would like to start there. But I agree that we should display the link as well when possible. That requires a little bit more work though. We need a way to display a waiting message and refresh from time to time to check if the server has generated the report already. Our new Cable technology could really help here.

* Step one with an email if the processing takes a lot of time: 2 days.
* Step two to also display a link when ready: + 3 days.
* Background processing plus link display without email: 4.5 days. :smile: 

That's just a rough guess because I'm not familiar with Cable yet.

**Comment by mkllnk** [comment]
@RachL Would you like us to reduce workers or increase memory in the meantime?

**Comment by RachL** [comment]
> Would you like us to reduce workers or increase memory in the meantime?

@mkllnk 😓 I don't know... what do you recommend?

Thanks for the estimation! Can we put the work behind a feature toggle? That would enable us to fine tune how to release it when we are close to finish: we start with the email, if we have more time we can add the notification within the page later (with a papercut we can add the ability for the user to turn off the email). Worse case scenario, first release big reports are accessible in email only and a later update will improve UX.

**Comment by abdellani** [comment]
@dacook @mkllnk 

I read the [article shared](https://scoutapm.com/blog/ruby-garbage-collection#alternativememory) by @dacook 

The base class of all the reports is `Reporting::ReportTemplate`.  If we have that circulate references issue, this report will not be collected.

So, I edited the `Reporting::ReportTemplate#initialize` as follows (I copied the code from the article)

```ruby
module Reporting
  class ReportTemplate
    def initialize(user, params = {}, request = nil)
      puts "--------------------------------"
      puts "initialize"
      puts "--------------------------------"
      ObjectSpace.define_finalizer(self, Reporting::ReportTemplate.finalizer_proc("abc"))
     ...
    end
    def self.finalizer_proc(name)

      proc { 
        puts "--------------------------------"
        puts "destroyed !!! "
        puts "--------------------------------"
      }
    end
  end 
end
```
I run the report `order_cycle_supplier_totals` for the first time, and the log contained only
```ruby
--------------------------------
initialize
--------------------------------
```

When I run the report multiple times, `finalizer_proc` is called from time to time.
```ruby
--------------------------------
destroyed !!! 
--------------------------------
```

I tried to remove the procs from  `OrderCycleSupplierTotals` and  `Base < ReportTemplate` but that didn't change much.

One conclusion: The Report object is not destroyed immediately after the request.

The thing that's confusing me now is why `finalizer_proc` is called from time to time when I refresh the report?

**Comment by abdellani** [comment]
I tried another test, this time I created circular references in the report template.

```ruby
    def initialize(user, params = {}, request = nil)
      puts "--------------------------------"
      puts "initialize"
      puts "--------------------------------"
      # ObjectSpace.define_finalizer(self, Reporting::ReportTemplate.finalizer_proc("abc"))
      ObjectSpace.define_finalizer(self, proc{
          puts "--------------------------------"
          puts " destroyed !!! "
          puts "--------------------------------"  
      })
...
    end

```

This time, I couldn't see `destroyed !!! "` regardless how many times I run the report.

This pushes me to say that we don't have circular references in the report because the garbage collector was able to destroy the object in the first test.

Also, from this [old discussion](https://www.ruby-forum.com/t/garbage-collection-and-define-finalizer/214342/2)

>You can’t rely on the GC calling your finalizers when the object is
>reaped. Each runtime will have different behavior.
 
My conclusion (please correct me if I'm wrong)

If the class has circular references, `finalize` will be called to process termination.
If the class doesn't have this issue, `finalize` may (or may not)  be called after responding to the client's request.

**Comment by mkllnk** [comment]
> we have that circulate references issue

It's not necessarily an issue. The controller references the report and the report references the controller. But at the end of the request, all variables of the controller are reset. The report is then not referenced by the controller any more. The garbage collector can then delete the report and with it the reference to the controller.

This is only happening from time to time because the garbage collector only runs from time to time.

> I tried another test, this time I created circular references in the report template.

That code from the article is *introducing* a circular dependency which is not cleaned up. It's an example of a problematic circular dependency. The finalizer is defined on the class which is not removed unless the class is destroyed. Unfortunately it's not destroyed though because a proc is referencing it via `self`. So there are two objects with permanent references to each other.

Our reports case is different because the `@report` ivar in the controller is not permanent. It's reset between requests and that removes the reference to the report.

There may be more references we haven't found yet but your test proves that at least the report object is clean up and we don't have a memory leak there. Otherwise we would see much bigger problems in production, I think. There may or may not be a small memory leak somewhere else but it could also just be the heap fragmentation filling up the memory slowly.

Anyway, the bigger problem at the moment is that just running big reports using more memory than we have. It doesn't matter if they are cleaned up afterwards. They use too much memory to start with.

**Comment by mkllnk** [comment]
I've been reading more about Puma, memory and connection management. I now think that increasing the Puma worker count for fr-prod was not helpful. If we want to handle more requests, we can simply increase the thread count which will use memory much more efficiently.
